From a10ee8a3006eb66599c61cdb20cb1f4053f4edd9 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 16 Jun 2018 16:23:38 -0400
Subject: [PATCH] Ignore Missing Recipes in RecipeBook to avoid data errors

This code was causing NPE's in saving player data, potentially related to reloads.

diff --git a/src/main/java/net/minecraft/server/RecipeBookServer.java b/src/main/java/net/minecraft/server/RecipeBookServer.java
index 121e1fd53..0e66bdda8 100644
--- a/src/main/java/net/minecraft/server/RecipeBookServer.java
+++ b/src/main/java/net/minecraft/server/RecipeBookServer.java
@@ -79,6 +79,10 @@ public class RecipeBookServer extends RecipeBook {
 
         while (iterator.hasNext()) {
             MinecraftKey minecraftkey = (MinecraftKey) iterator.next();
+            // Paper start - ignore missing recipes
+            final Optional<? extends IRecipe<?>> recipe = this.l.a(minecraftkey);
+            if (!recipe.isPresent()) continue;
+            // Paper end
 
             nbttaglist.add(new NBTTagString(minecraftkey.toString()));
         }
@@ -89,6 +93,10 @@ public class RecipeBookServer extends RecipeBook {
 
         while (iterator1.hasNext()) {
             MinecraftKey minecraftkey1 = (MinecraftKey) iterator1.next();
+            // Paper start - ignore missing recipes
+            final Optional<? extends IRecipe<?>> recipe = this.l.a(minecraftkey1);
+            if (!recipe.isPresent()) continue;
+            // Paper end
 
             nbttaglist1.add(new NBTTagString(minecraftkey1.toString()));
         }
-- 
2.22.0

